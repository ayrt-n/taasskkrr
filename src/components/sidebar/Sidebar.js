import React, { useState, useEffect} from 'react';
import '../../styles/Sidebar.css'
import SidebarItem from './SidebarItem'
import SidebarCollapsableList from './SidebarCollapsableList';
import UserService from '../../services/UserService';
import ProjectModal from '../tasks/ProjectModal';
import NewProjectButton from './NewProjectButton';

function Sidebar({ userProjects, addProject }) {
  const [modal, setModal] = useState({isOpen: false, action: '', data: {}});
  const inbox = userProjects.filter((project) => (project.inbox))[0];
  const projects = userProjects.filter((project) => (!project.inbox));

  const openModal = (action, data) => {
    setModal(
      {
        isOpen: true,
        action: action,
        data: data
      }
    )
  };

  const closeModal = () => {
    setModal(
      {
        ...modal,
        isOpen: false,
      }
    )
  }

  return (
    <div className="Sidebar">
      <ProjectModal
        action={modal.action}
        data={modal.data}
        isOpen={modal.isOpen}
        closeModal={closeModal}
      />
      <SidebarItem icon="inbox.svg" title="Inbox" action={`/projects/${inbox.id}`} />
      <SidebarItem icon="today.svg" title="Today" action="/today" />
      <SidebarItem icon="upcoming.svg" title="Upcoming" action="/upcoming" />
      <SidebarCollapsableList title="Projects" items={projects} handleAdd={addProject} />
      <NewProjectButton openModal={openModal} afterSubmit={addProject} />
    </div>
  );
}

export default Sidebar;
