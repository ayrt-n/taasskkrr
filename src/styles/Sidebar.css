.Sidebar {
  box-sizing: border-box;
  overflow-y: auto;
  width: 280px;
  position: fixed;
  top: 60px;
  bottom: 0;
  padding: 2rem 1.5rem;
  background-color: var(--bgColor);
  z-index: 1;

  border-top: 2px solid var(--bgLayerColor);;
  border-right: 2px solid var(--bgLayerColor);
}

.SidebarItem {
  display: flex;
  align-items: center;
  cursor: pointer;
  padding: 0.75rem;
  border-radius: 3px;
  
  /* CSS to prevent selecting text */
  -webkit-touch-callout: none; /* iOS Safari */
  -webkit-user-select: none; /* Safari */
  -khtml-user-select: none; /* Konqueror HTML */
  -moz-user-select: none; /* Old versions of Firefox */
  -ms-user-select: none; /* Internet Explorer/Edge */
  user-select: none; /* Non-prefixed version, currently */
}

.SidebarItem.list-header {
  justify-content: space-between;
}

.SidebarItem-button {
  background: none;
  border: none;
  color: var(--bgLightLayerColor);
  display: flex;
  align-items: center;
  cursor: pointer;
  transition: color 150ms;
  font-size: 1rem;
}

.SidebarItem-button:hover {
  color: var(--fgColor);
}

.SidebarItem-button:active {
  color: var(--bgLayerColor);
}

.SidebarItem.no-hover:hover {
  background-color: inherit;
  color: inherit;
}

.SidebarItem.SubItem .SidebarItem-icon {
  margin-left: 1.125rem;
}

.SidebarItem:not(:last-child) {
  margin-bottom: 0.25rem;
}

.SidebarItem:hover {
  background-color: var(--bgLayerColor);
  color: #FFFFFF;
}

.SidebarItem-icon {
  width: 1.5rem;
  height: auto;
  margin-right: 0.75rem;
}

.SidebarItem-title {
  font-weight: 600;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
}

.collapsed .Sidebar-collapsable-items {
  max-height: 0;
  overflow: hidden;
}
