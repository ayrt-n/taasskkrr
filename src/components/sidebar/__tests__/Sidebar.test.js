import React from 'react';
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import Sidebar from '../Sidebar';

// Mock child components
jest.mock('../SidebarItem', () => ({ title }) => (
  <div data-testId="item">{title}</div>
));

jest.mock('../SidebarCollapsableList', () => ({ children }) => (
  <div>{ children }</div>
));

jest.mock('../NewProjectButton', () => ({ openModal }) => (
  <button onClick={openModal}>Add Project</button>
));

// Set up props
const inbox = { id: 1 };
const projects = [
  { title: 'Project 1', id: 2 },
  { title: 'Project 2', id: 3 },
];
const addProjectMock = jest.fn();
const openModalMock = jest.fn();


describe('Sidebar component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Sidebar
        projects={projects}
        inbox={inbox}
        addProject={addProjectMock}
        openModal={openModalMock}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders all sidebar items and projects', () => {
    render(
      <Sidebar
        projects={projects}
        inbox={inbox}
        addProject={addProjectMock}
        openModal={openModalMock}
      />
    );

    const sidebarItems = screen.queryAllByTestId('item')
    expect(sidebarItems.length).toBe(5);
  });
});
