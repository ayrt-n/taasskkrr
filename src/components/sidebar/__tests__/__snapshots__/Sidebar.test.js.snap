// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar component renders correctly 1`] = `
<div>
  <div
    class="Sidebar"
  >
    <div
      data-testid="item"
    >
      Inbox
    </div>
    <div
      data-testid="item"
    >
      Today
    </div>
    <div
      data-testid="item"
    >
      Upcoming
    </div>
    <div>
      <div
        data-testid="item"
      >
        Project 1
      </div>
      <div
        data-testid="item"
      >
        Project 2
      </div>
    </div>
    <button>
      Add Project
    </button>
  </div>
</div>
`;
